File: sample/src/main/java/ru/surfstudio/standard/ui/base/recycler/PaginationableAdapter.java
package ru.surfstudio.standard.ui.base.recycler;

import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import me.zhanghai.android.materialprogressbar.MaterialProgressBar;
import ru.surfstudio.android.core.ui.base.recycler.BasePaginationableAdapter;
import ru.surfstudio.android.core.ui.base.screen.model.state.PaginationState;
import ru.surfstudio.standard.R;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

/**
 * адаптер для пагинируемых списков
 */
public class PaginationableAdapter extends BasePaginationableAdapter {

    private PaginationFooterItemController paginationFooterItemController;

    @Override
    protected BasePaginationFooterController getPaginationFooterController() {
        if (paginationFooterItemController == null) {
            paginationFooterItemController = new PaginationFooterItemController();
        }
        return paginationFooterItemController;
    }

    protected static class PaginationFooterItemController extends BasePaginationFooterController<PaginationFooterItemController.Holder> {

        @Override
        protected Holder createViewHolder(ViewGroup parent, OnShowMoreListener listener) {
            return new Holder(parent, listener);
        }

        protected class Holder extends BasePaginationFooterHolder {

            private ProgressBar progressBar;
            private TextView showMoreTv;

            public Holder(ViewGroup parent, OnShowMoreListener listener) {
                super(parent, R.layout.pagination_footer_layout);
                showMoreTv = itemView.findViewById(R.id.pagination_footer_text);
                progressBar = (MaterialProgressBar) itemView.findViewById(R.id.pagination_footer_progress);
                showMoreTv.setOnClickListener(v -> listener.onShowMore());
                progressBar.setVisibility(GONE);
                showMoreTv.setVisibility(GONE);
            }

            @Override
            public void bind(PaginationState state) {
                switch (state) {
                    case LOADING:
                        progressBar.setVisibility(VISIBLE);
                        showMoreTv.setVisibility(GONE);
                        break;
                    case COMPLETE:
                    case READY:
                        progressBar.setVisibility(GONE);
                        showMoreTv.setVisibility(GONE);
                        break;
                    case ERROR:
                        progressBar.setVisibility(GONE);
                        showMoreTv.setVisibility(VISIBLE);
                        break;
                    default:
                        throw new IllegalArgumentException("unsupported state: " + state);
                }
            }
        }
    }
}

File: template/src/main/java/ru/surfstudio/standard/ui/base/recycler/PaginationableAdapter.kt
package ru.surfstudio.standard.ui.base.recycler

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.view.animation.AccelerateInterpolator
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.view.updateLayoutParams
import ru.surfstudio.android.easyadapter.pagination.BasePaginationableAdapter
import ru.surfstudio.android.easyadapter.pagination.PaginationState
import ru.surfstudio.standard.R

class PaginationableAdapter : BasePaginationableAdapter() {

    private var paginationFooterItemController: BasePaginationFooterController<*>? = null


    override fun getPaginationFooterController(): BasePaginationFooterController<*> {
        if (paginationFooterItemController == null) {
            paginationFooterItemController = PaginationFooterItemController()
        }
        return paginationFooterItemController!!
    }

    private class PaginationFooterItemController : BasePaginationFooterController<PaginationFooterItemController.Holder>() {

        override fun createViewHolder(parent: ViewGroup, listener: BasePaginationableAdapter.OnShowMoreListener): Holder {
            return Holder(parent, listener)
        }

        inner class Holder(
                parent: ViewGroup,
                listener: BasePaginationableAdapter.OnShowMoreListener
        ) : BasePaginationFooterHolder(parent, R.layout.pagination_footer_layout) {

            private val progressBar: ProgressBar
            private val showMoreTv: TextView

            init {
                showMoreTv = itemView.findViewById<TextView>(R.id.pagination_footer_text)
                progressBar = itemView.findViewById<ProgressBar>(R.id.pagination_footer_progress)
                showMoreTv.setOnClickListener { _ -> listener.onShowMore() }
                progressBar.visibility = GONE
                showMoreTv.visibility = GONE
            }

            override fun bind(state: PaginationState) {

                when (state) {
                    PaginationState.READY -> {
                        progressBar.visibility = VISIBLE
                        showMoreTv.visibility = GONE
                    }
                    PaginationState.COMPLETE -> {
                        progressBar.visibility = GONE
                        showMoreTv.visibility = GONE
                    }
                    PaginationState.ERROR -> {
                        progressBar.visibility = GONE
                        showMoreTv.visibility = VISIBLE
                    }
                    else -> throw IllegalArgumentException("Unsupported state: " + state)
                }
            }

            override fun animateRemove(): Boolean {
                itemView.animate()
                        .alpha(0f)
                        .setInterpolator(AccelerateInterpolator())
                        .setListener(object : AnimatorListenerAdapter() {
                            override fun onAnimationEnd(animation: Animator?) {
                                itemView.alpha = 1f
                            }

                            override fun onAnimationCancel(animation: Animator?) {
                                itemView.alpha = 1f
                            }
                        })
                        .start()
                return false
            }


        }
    }
}